pkg/controller/daemon/daemon_controller.go

```diff
@@ -223,7 +223,7 @@ func indexByPodNodeName(obj interface{}) ([]string, error) {
 
 func (dsc *DaemonSetsController) addDaemonset(obj interface{}) {
 	ds := obj.(*apps.DaemonSet)
-	klog.V(4).Infof("Adding daemon set %s", ds.Name)
+	klog.V(4).InfoS("Adding DaemonSet", "daemonset", klog.KObj(ds))
```

Comment:
what is the performance impact of calling KObj for constructing parameters to trace logging calls? KObj does reflection and allocation on every call, even when the results will not be used at all (like when we're logging below V(4) trace levels)

Reply:
I'm not certain but I believe the go compiler elides the heap allocation in ValueOf in `klog.KObj` and keeps it on the stack.  I wouldn't be worried about it unless it's showing up in a profile in most components, although controllers definitely get the most calls in tight loops to logging.

Reply:
It inlines the call, but doesn't elide heap allocations for reflect.  And as Jordan notes Go calling convention will always execute KObj (in case it has side effects) unless you put the explicit if.

I'm kind of surprised the heap allocation isn't elided.

Reply:
Actually, I was wrong, the heap allocation IS elided, so `klog.KObj` with go 1.16 in this fashion doesn't allocate.  But InfoS *has* to allocate to convert KObj to an interface{} to fit in the slice, so the outcome is the same:

```
func TestKlog(t *testing.T) {
	pod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo"}}
	b := klog.V(4)
	for i := 0; i < 1000000; i++ {
		a := klog.KObj(pod)
		b.InfoS("Blah", a)
	}
}

```

```
$ go test -o /tmp/blah ./pkg/kubelet/types -test.run=TestKlog -cpuprofile /tmp/cpu -memprofile /tmp/mem
$ go tool pprof -alloc_objects -http : /tmp/blah /tmp/mem
```

Go to "Source" in pprof, you can see the allocations are in the invocation of the line for the klog.  However, I think the reason there is an allocation is not the reflect, but the fact that you're passing a stack value to the InfoS []interface{} argument, which requires a new object to be allocated on the heap.  So the allocation is a side effect of InfoS accepting `...interface{}`.

So in hotter loops as jordan notes yes you should use the `if klog.V(4).Enabled() { klog.InfoS ... }` form.

Reply:
@smarterclayton I'm not sure what action I should take here. Should I use the if statement you mentioned above? If so, how do I know what is a "hotter loop"?